### PR TITLE
ENH: Change COO index dtype so reshape will not fail

### LINKED ISSUE #8912: Reshaping COO sparse matrices fails if the matrix has 2^31 or more elements
The [`coo_matrix.reshape` method](https://github.com/scipy/scipy/blob/master/scipy/sparse/coo.py#L194-L224) computes the linear index of each nonzero element (variable `flat_indexes` in the code) as an intermediate step. If there are more than 2^31 elements, the linear index overflows to negative and then the construction of the new matrix blows up. (This is the number of elements in the abstract matrix, not the number of nonzero elements in the sparse representation.)

This is particularly problematic because COO reshape is the default reshape method for sparse matrices so it affect CSC, CSR, and others.

What do you think the best solution is?

The obvious path is to do nothing and require that users supply the indexes as int64 if they want correct reshape behavior even if int32 is sufficient for the row and column indexes. However, this is very painful. Constructing CSR and CSC matrices with int64 indexes is not possible. If the user provides the rows and columns or indices and indptr as int64, the indexes and indptr get lowered to int32 anyway.

There are a couple of conservative solutions to the problem:

1. Raise the dtype to np.int64 for the linear index calculation.
2. Raise the dtype to np.int64 for the linear index calculation only when then number of elements is 2^31 or greater.

In the first two cases, we would have to decide under what circumstances we should lower the dtype back down to int32. 

It might be possible to rewrite the index calculation so that it simply does not overflow, but it is not obvious to me how to do that.

### PR DESCRIPTION
**Pull request by drhagen**
This is a new attempt to fix #8912 after my first attempt in #8983 failed.

### CODE REVIEW

File: scipy/sparse/coo.py
Comment (tylerjereddy) [inline scipy/sparse/coo.py]: It looks like `get_index_dtype` was always casting to `np.int32` if the largest dimension in the array didn't overflow single-precision int storage for index generation. The new behavior encoded in this PR is that a set of indices over **all** the elements of an array (not just the rows or columns in isolation) has to fit in single precision int, otherwise `np.int64` will be used.

If I understand the linked issue discussion, the idea is to prevent a downstream sparse reshape operation from receiving an int32 dtype, which was working ok for the row / column indexing, but won't work for a reshape where the indices of all elements in the array may not fit in int32?

Comment (tylerjereddy) [reply scipy/sparse/coo.py]: `self.size` may be more concise (& clear?) than manually calculating the product here?

**Review by tylerjereddy — COMMENTED** [review]
The CI failures appear to be related to changes, but there are just a smaller number of them in Travis.

I'm guessing there's no cleaner way to patch this by just using the `size` for the reshape operation since you'll still have to deal with up or down-casting as appropriate?

File: scipy/sparse/coo.py
Comment (drhagen) [reply scipy/sparse/coo.py]: > The new behavior encoded in this PR is that a set of indices over all the elements of an array (not just the rows or columns in isolation) has to fit in single precision int, otherwise `np.int64` will be used.

This is exactly correct.

> `self.size` may be more concise (& clear?) than manually calculating the product here?

You would think that `self.size` is the number of elements in the matrix, like `numpy.matrix`, but [it's not](https://github.com/scipy/scipy/blob/v1.1.0/scipy/sparse/base.py#L683-L684). It's the number of nonzero elements in the sparse array. This multiplication is the only way until we deprecate `size` and require people to use `nnz`, instead.

**Comment by drhagen** [comment]
> The CI failures appear to be related to changes

I agree. I could use some advice on how to fix these failures. I am trying to figure out what `Test64Bit` does exactly. The test I wrote passes, but it run a few more times by this class where it fails.

**Comment by drhagen** [comment]
I think the remaining failure on Travis is spurious and that this PR is ready for review.

File: scipy/sparse/tests/test_base.py
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py]: Not the number of elements, really, but the maximum linear index.

File: scipy/sparse/coo.py
Comment (perimosocordiae) [inline scipy/sparse/coo.py]: This will cause additional memory usage for some users. That may or may not be an issue, but it's worth mentioning.

I may have missed it, but did you try an on-demand approach inside the `reshape` method? That is, detecting the overflow case and promoting to `int64` there?

File: scipy/sparse/tests/test_base.py
Comment (drhagen) [reply scipy/sparse/tests/test_base.py]: I've always envisioned a 4x3 sparse matrix as still being a matrix with 12 elements; it's just that only the nonzero elements take memory. ;-) I'll clarify the comment because that is apparently not universal.

File: scipy/sparse/coo.py
Comment (drhagen) [reply scipy/sparse/coo.py]: Yes, in the case that both dimensions are less than 2^31 but the maximum linear index is greater than 2^31, this will cause the matrix to require more memory than strictly necessary.

I actually first tried it the way you suggested (upcast only in `reshape`), but it was not a clean solution, so I switched to this solution. The biggest issue that if you upcast at the beginning of `reshape`, then you need to downcast at the end (but only when the dims are still both less than 2^31). If you upcast in the constructor, then you don't need to do anything because `reshape` always preserves the maximum linear index. There is a time-space tradeoff, where upcasting and downcasting can save memory by burning extra cycles on the casting when needed.

I am open to switching back to the upcasting and downcasting technique if saving memory in all cases is paramount.

Comment (perimosocordiae) [reply scipy/sparse/coo.py]: How would you feel about making the constructor less insistent on using the dtype given by `get_index_dtype`? That is, if the user provides index arrays with dtype larger or equal to the minimum dtype, just use them as-is.

```python
minimal_idx_dtype = get_index_dtype(maxval=max(self.shape))
idx_dtype = np.find_common_dtype([row.dtype, col.dtype, minimal_idx_dtype], [])
```

This way we could upcast on demand when reshaping, without having to downcast later.

(Note that my code snippet isn't quite sufficient because `row` and `col` might not be numpy arrays with dtypes, but the idea is the same.)

Comment (drhagen) [reply scipy/sparse/coo.py]: I would find it strange that a matrix with 32-bit indexes becomes a matrix with 64-bit indexes after reshaping, but only sometimes. `reshape` returns `self` if the shape is unchanged, so the upcast would not happen in that case, presumably. It really depends on how predictable you want the dtype behavior to be.

Comment (perimosocordiae) [reply scipy/sparse/coo.py]: Fair point. Let's merge this as-is, then if anyone complains we can explore other options.

**Review by perimosocordiae — APPROVED** [review]
I'm +1 to merge. @tylerjereddy any further comments?

If there's no motion for a few days, I'll merge then.

**Review by tylerjereddy — APPROVED** [review]
+1 from me now as well -- I'll let @perimosocordiae hit the merge though since sparse is well outside my domain of expertise :)

**Comment by tylerjereddy** [comment]
The mac failure is not related

**Comment by perimosocordiae** [comment]
Merged, thanks @drhagen and @tylerjereddy !

**Comment by pv** [comment]
I would have here preferred `coo_matrix.reshape` to recompute `idx_dtype` of the index arrays, and do intermediate calculations with the larger integer type. I think it's not unexpected that the index dtype may change in reshape operation.

**Pull request merged**
